lib/IRGen/StackSizeLimitTransform.cpp

```diff
@@ -0,0 +1,307 @@
+#include "swift/AST/IRGenOptions.h"
+#include "swift/Basic/BlotMapVector.h"
+#include "swift/Basic/FrozenMultiMap.h"
+#include "swift/IRGen/IRGenSILPasses.h"
+#include "swift/SIL/InstructionUtils.h"
+#include "swift/SIL/SILBasicBlock.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILInstructionWorklist.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/CanonicalizeInstruction.h"
+#include "swift/SILOptimizer/Utils/InstOptUtils.h"
+#include "llvm/IR/Constants.h"
+
+using namespace swift;
+using namespace swift::irgen;
+
+//===----------------------------------------------------------------------===//
+//                                  Utility
+//===----------------------------------------------------------------------===//
+
+/// Given an alloc_stack \p asi and the next element to process, convert asi's
+/// users to use a new heap allocated box without invalidating next.
+static SILBasicBlock::iterator
+convertStackBoxToHeapBox(AllocStackInst *asi, SILBasicBlock::iterator next) {
+  SILBuilderWithScope builder(asi);
+  auto *box =
+      builder.createAllocBox(asi->getLoc(), asi->getType().getObjectType());
+  auto *projectBox = builder.createProjectBox(asi->getLoc(), box, 0);
+  while (!asi->use_empty()) {
+    auto *use = *asi->use_begin();
+
+    // Replace all of the alloc_stack's dealloc_stack with a destroy_value of
+    // the box.
+    if (auto *dsi = dyn_cast<DeallocStackInst>(use->getUser())) {
+      SILBuilderWithScope destroyBuilder(dsi);
+      destroyBuilder.emitDestroyValueOperation(dsi->getLoc(), box);
+      if (next == dsi->getIterator())
+        ++next;
+      dsi->eraseFromParent();
+      continue;
+    }
+
+    // Otherwise, make the alloc_stack user now use the project_box.
+    use->set(projectBox);
+  }
+
+  // Then erase the alloc_stack, incrementing next as we need to.
+  if (next == asi->getIterator())
+    ++next;
+  asi->eraseFromParent();
+  return next;
+}
+
+//===----------------------------------------------------------------------===//
+//                       Per Instruction Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct StackSizeLimitState {
+  SmallVector<std::pair<AllocRefInst *, BuiltinInst *>, 32> builtinsToProcess;
+  bool madeChange = false;
+  int stackAllocSize;
+  IRGenModule &igm;
+
+  // Max stack promotion that we can perform in a single function.
+  StackSizeLimitState(IRGenModule *igm)
+      : stackAllocSize(igm->IRGen.Opts.StackPromotionSizeLimit), igm(*igm) {}
+};
+
+} // anonymous namespace
+
+static SILBasicBlock::iterator
+processAllocStackInst(StackSizeLimitState &state, AllocStackInst *asi,
+                      SILBasicBlock::iterator next) {
+  // First see if we have already hit our size limit. If we have, just
+  // convert to a heap allocation.
```

Comment:
I don't think it's a good idea to convert stack allocations to heap allocations. AFAIK, IRGen doesn't do that. It would introduce hidden performance cliffs which we should try to avoid.

Reply:
The reason I was playing with this is that I have heard from various stdlib people that being able to move alloc_stack onto the heap for large alloc_stack was desirable. My memory could be wrong. @Catfish-Man.

One thought is that we could have a different budget for alloc_stack or have a mode where this behavior is enabled.

Reply:
(One reason we don't encounter this problem today is that *all* stack promotion for Arrays is broken)

Reply:
@Catfish-Man I'm not talking about stack promotion. I was referring to stack allocations which are on the stack by default, like small value types, etc. I don't think the compiler should _introduce_ new heap allocations.
For stack promotion, it's a different story, because by default classes/array-buffers are heap allocated. So the user would expect heap allocation here.

> all stack promotion for Arrays is broken

That's not true

Reply:
@eeckstein I think it is broken. Try this example:

```
@inline(never)
func myF(_ i: Int) {
    var s = [i,2,3]
    s.append(4)
    s.append(5)
    print(s)
}

myF(5)
```

Reply:
We don't stack allocate arrays with a non-fixed size.

Reply:
In my tests this allows for a 4x speedup on quite a few benchmarks. It's possible that there's some other explanation, but… it seems unlikely. Either way, we should take 4x speedups when we can get them.